### PR TITLE
Server channel affinity

### PR DESCRIPTION
**Pull request by sreecha**
Pull request for https://github.com/grpc/grpc/issues/5470

### CODE REVIEW

**Comment by sreecha** [comment]
The following change in `server.c` seems to be causing all the failures in `Basic tests`
https://github.com/grpc/grpc/pull/6149/files#diff-e349c3f11efda23128ed77cc156dc260L1036

I am working on fixing the tests..

**Comment by sreecha** [comment]
The hybrid end2end tests failed. The tests will have to be rewritten . The following is the issue:

The problem here is that the sync server uses a `ServerCompletionQueue` (lets call it `cq1` ) and ensures that an internal thread always calls `grpc_completion_queue_next()` on `cq1`. (Note that even if the application registers more completion queues with the server, the Sync server does not call `grpc_completion_queue_next()` on those completion queues)

Now, when using the hybrid model, the async part of the server implementation (which is owned by the application implementing the async server) will have to create a new completion queue (lets call it `cq2`) and register with the server. To handle Async methods, it will have to do `grpc_completion_queue_next()` on `cq2` [note: It has to create a new completion queue since it cannot access the Sync server's completion queue `cq1` - i.e there is no API that exposes that server completion queue] 

When the server starts:
-- The C++ sync server part ensures that `grpc_completion_queue_next()` is called on `cq1` on thread `T1`
-- The application calls `grpc_completion_queue_next()` on `cq2` on thread `T2`

A client creates a channel to this server. Both `T1` and `T2` wake up. If we assume that thread `T2` is the one that actually `accepts` the channel , then as per this PR, the channel is affinitized to `cq2` (i.e the new `fd` for the channel is added to the pollset corresponding to `cq2`).

Now all calls on that channel will only be processed if `grpc_completion_queue_next()` is called regularly on `cq2`. If not, the calls would stall (from client's perspective). This is what was happening in the hybrid tests.

**Comment by vjpai** [comment]
It seems inappropriate to rewrite those tests to workaround this issue as the PR is a performance optimization, not an API change. The API that those tests were written under is still valid.

**Comment by ctiller** [comment]
I think Vijay's right. I missed this case. Still, this optimization is
important, and hybrid mode is important. Let's put some thought into core
API changes that could make this work?

An initial thought: what if we passed some group identifier down with
register_completion_queue and affinitized one cq per group?

**Comment by sreecha** [comment]
Thanks for reading my update :) and quickly responding Vijay and Craig!..

Looks like misunderstood the contract of Async server implementations (I am assuming whats true for Async would be true for Hybrid servers too).  If the application 'registers' multiple 'server' completion queues, isn't the expectation that the application is responsible to ensure that it is calling `grpc_completion_queue_next` (via `cq->Next()` or `cq->AsyncNext()`) regularly on those completion queues ?

Or is it okay for it to just call `cq->Next()` a couple of times then just ignore it until server shutdown ? (which is what the tests are doing).

**Comment by ctiller** [comment]
Ah... actually you might have a point (with further explanation).

For async servers at least there are two common patterns:
- a cq is registered and is entered frequently
- a cq is registered and is entered never (until server shutdown)

I think I'm ok with the restriction (and we should write this up in the
ServerBuilder docs) that once you start calling CQ.Next(), you need to
continue calling it relatively frequently.

Perhaps a bit on a completion queue when registering it would make sense
also? Indicating that this CQ is OK to use for listening to new
connections? (we'd probably under document it and let super advanced users
only know about it).

**Comment by vjpai** [comment]
I like the bit idea.

**Comment by sreecha** [comment]
Cool. Thanks for more details.

I too like the idea of adding a 'bit' on the completion queue to say it is ok to use for listening to new connections.  Looks like this `bit` would be on by default...I will think about this a bit  :) more to see if there is any good way to avoid breaking any existing clients..

**Comment by sreecha** [comment]
Added the option to mark a server completion queue as 'non listening'.  A non listening completion queue is not added to server's cq list (and hence does not get any listening fd in its pollset). 

I am using the term "is_frequently_polled" at the C++ layer (and "non_listening" at the core layer). I felt that the term `is_frequently_polled` would clearer than `is_non_listening` to the user calling `ServerBuilder::AddCompletionQueue(bool is_frequently_polled = true)`  API

File: include/grpc++/server_builder.h [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- * Copyright 2015, Google Inc.
+ * Copyright 2015-2016, Google Inc.
```
Comment (ctiller) [inline include/grpc++/server_builder.h]: No need to change the copyright notice

File: include/grpc/grpc.h [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- * Copyright 2015, Google Inc.
+ * Copyright 2015-2016, Google Inc.
```
Comment (ctiller) [inline include/grpc/grpc.h]: No need

**Comment by sreecha** [comment]
Merged @ctiller 's change cefa37803acfd59071d3e7bee2bcf8537c3ec8ae to add affinity to the new `ev_poll_posix.c`.

**Comment by jtattermusch** [comment]
Tests don't look too bad. Probably some finishing touch is needed but overall it looks not terribly broken.
https://grpc-testing.appspot.com/job/gRPC_pull_requests/9063/

**Comment by sreecha** [comment]
This change is subsumed by https://github.com/grpc/grpc/pull/6612

**Pull request merged**